Review CMake support in magrit client

daniperez opened the pull request:
Could you test this?

It's working in Fedora but I need the Ubuntu mingw details. Have a look at Windows-mingw-i686.cmake, on top. I need the MINGW_PREFIX and Boost_COMPILER. The former is used to name the compiler itself (e.g.: ${MINGW_PREFIX}-g++ or i686-pc-mingw32-g++ in my case) and the latter is added to the library name (e.g.: libboost_program_options${Boost_COMPILER}-1_46_1.dll.a or libboost_program_options-gcc45-1_46_1.dll.a in my case).

The previous two variables should be enough but it may be the case that you had also to supply the CMAKE_FIND_ROOT_PATH in Ubuntu. I'll be happy enough if it builds in my Fedora and your Ubuntu installations.

With regards to maven, I'm just building for Linux. Let's discuss how we do to build all the platforms (the three of them serially? depending on a env. variable? or platform auto-detect?).

daniperez committed:
First commit.

daniperez committed:
--help/--version working

daniperez committed:
magrit-build almost done

daniperez committed:
program_options somehow working

DPerez-Voxel committed:
compilation fixed... thinking about dropping boost::program_options support

DPerez-Voxel committed:
Removing boost's shared_ptr, using stl's.

Couldn't get rid of for_each, C++11 for not supported by my compiler
yet.

DPerez-Voxel committed:
being more precise :-)

daniperez committed:
generic_command pop out of utils!

daniperez committed:
Help notice improved

daniperez committed:
Started processing positional parameters.

daniperez committed:
Started using .cpp's. Added a join() specialization for strings,
I couldn't make use of the generic method with iterators :-(

daniperez committed:
Makefile enhanced

daniperez committed:
Renaming magrit-build. Using vector<string> instead
of C arrays.

daniperez committed:
fix of arguments vector.

daniperez committed:
interface of get_subcommands changed to avoid returning iterators
to temporary vectors.

daniperez committed:
Removed uncompleted commands, for menthal sanity :-)

daniperez committed:
Permissions unified

daniperez committed:
Added a description for each subcommand.

Not yet working as I wanted ("build" command appearing
regardless of the subcommand we are "--help"-ing ).

daniperez committed:
Added the rest of commands. They're not implemented though.

daniperez committed:
turning generic_command into a class to make it more clear.

daniperez committed:
--help support working (yet to polish output and chain subcommands)

daniperez committed:
Command line parsing fixed. Started using a namespace after some clashes.

daniperez committed:
commands look better now

daniperez committed:
Command line issues solved

daniperez committed:
enhancement to deal with "foo --command bar"/"foo bar" extreme case (description
in the source)..

daniperez committed:
Showing the full command in help

daniperez committed:
Added a description to the command being requested

daniperez committed:
minor

daniperez committed:
Order of match changed.

We try to match first the line as is
and only if it didn't mathc, we try to match a subcommand.

daniperez committed:
Error output enhanced

daniperez committed:
Started implementing build-log.

daniperez committed:
Started using the pipes

daniperez committed:
more work on build log

daniperez committed:
aesthetic change to positional parameter mgmnt. do_not_continue() becomes success().

daniperez committed:
More work on config*

daniperez committed:
Producing a binary for magrit-build-log.

We'll replacing bash implementation command by command, first
magrit-build-log.

daniperez committed:
first try

daniperez committed:
Using fork instead popen

daniperez committed:
Arguments to exec were wrong

daniperez committed:
no much progress. I have to switch to boost::process for portability and
menthal sanity.

Commit:
Fix for 64bits

Commit:
trying boost::process

Commit:
makefile fixed

Commit:
First compiling version of boost::process

daniperez committed:
minor

daniperez committed:
Merge branch 'master' of github.com:daniperez/magrit

daniperez committed:
start_process with boost::process::launch works.

daniperez committed:
Almost there :-) git log streamed to ssh command. That's fast!

(not fully working yet)

daniperez committed:
Basic build-log working.

I have to cleanup things, but the basics are there. boost::process
seems to do the trick despite I cannot set pipelines as I was
expecting to.

daniperez committed:
Merge branch 'boost-process'

Conflicts:
	client/src/main/c++/Makefile

daniperez committed:
repo name, user and host implemented.

daniperez committed:
Fix for some widespread linux distros shipping gcc < 4.6

daniperez committed:
Re-implemented with Boost Process' pipeline.

daniperez committed:
pipeline code re-factorized.

daniperez committed:
Status colors implemented.

ptitfred committed:
C++ build integration

'mvn package' to produce the client zip package

DPerez-Voxel committed:
Merging .gitignores and a added a "todo" note for later.

daniperez committed:
Minor change.

daniperez committed:
Added git_args to the second "git log" issued. Some code removed.

"I love the smell of removed code in the evening..."

daniperez committed:
Faster build, faster execution (maybe) and smaller executable (yet big though...)

daniperez committed:
Added color/debug options. Removed -O3 hoping it will fix compilation issues
in old compilers.

daniperez committed:
minor

ptitfred committed:
Fix maven integration

daniperez committed:
Merge pull request #2 from ptitfred/for-dani

Fix maven integration

daniperez committed:
Color fixed and minor change in pipeline's error.

daniperez committed:
Coloring up a bit the errors and throwing an error on options not
recognized (I was already throwing an error for non-recognized commands
but not options with dashes).

daniperez committed:
Merge branch 'master' of github.com:daniperez/magrit

Conflicts:
	client/src/main/c++/Makefile

daniperez committed:
minor

ptitfred committed:
Merge remote-tracking branch 'daniperez/master' into for-dani

daniperez committed:
Merge pull request #5 from ptitfred/for-dani

Robot first draft - includes a test of magrit build-log

daniperez committed:
Program options as positional options are now allowed.

Nevertheless, there's still one case to implement that it's not yet
considered. When using the magrit executable (magrit), if
an unrecognized options is passed *before* the positional parameters,
it will be accepted. E.g.:

  magrit build -1 log

will be considered the same as:

  magrit build log -1

We are not using the magrit executable, but just in case.

daniperez committed:
Color working but it's changing the width of the output.

daniperez committed:
This fixes #3: log messages cut down to a given size.
Configurable with git's variable magrit.log.maxWidth.

daniperez committed:
removed space between ellipsis and log message

daniperez committed:
Added wait.?pp and creating and intermediate library.

daniperez committed:
forgot this

daniperez committed:
Message width takes now into account sha1 signature and spaces.

ptitfred committed:
Merge branch 'robot' into for-dani

daniperez committed:
Merge pull request #6 from ptitfred/for-dani

Robot update

daniperez committed:
Getting prepared to compile for other platforms.

By getting stricter with compilation warnings, some porting bugs
can be early detected (e.g. typical bug is to use an integer
that gets truncated in win32 but not in linux64. It can be difficult
to spot once the code is written.  -Wconversion avoids that).

I had to fix very few lines of boost-process to stick to the new
stricter rules.

daniperez committed:
Merge branch 'master' of github.com:daniperez/magrit

daniperez committed:
Succeeded to compile for win32 in Fedora 15.

Yet to decide if:

 i)  shipping the needed dlls (it's ugly: even libstdc++ is
     not found. Does such a dll exist in Windows??))
 ii) or using static linking. I didn't succeed to create
     a statically linked libmagrit.dll.

daniperez committed:
Started implementing wait-for. Some warnings fixed.

daniperez committed:
This should fix a warning only reproducible in Ubuntu.

daniperez committed:
This should fix the compilation warnings in 64bits platforms.

daniperez committed:
wait-for code more or less finished. Not working yet!

This commit can cause severe damage to your computer.

daniperez committed:
Logging message added while waiting for events.

daniperez committed:
Started implementing --watch option. print_status re-factored in order to use it with the --watch option.

daniperez committed:
--watch implemented, not working though :-) but almost there!

daniperez committed:
print_date implemented. Removed stdout for magrit-wait-for.

daniperez committed:
--watch more or less working.

ptitfred committed:
Fix linking of Boost (at least under Ubuntu64)

daniperez committed:
Merge pull request #8 from ptitfred/fix/c++

Fixes a linking order problem reproducible in Ubuntu.

ptitfred committed:
MAGRIT_VERSION passed to C++ compilation

daniperez committed:
Started implementing CMake support. So far it is in works-in-my-desktop state.

Also added some files that we will need to provide if the client is packaged.

If you want to give it a try:

 mkdir build
 cd build
 cmake -C ../conf/linux32.conf ../ # The "-C <conf>" flag is not used yet, it can be omitted
 make

To note that the CMake script is not as evolved as the current Makefile, this is
just work in progress.

daniperez committed:
Merge pull request #9 from ptitfred/for-dani

danke!

daniperez committed:
Merge branch 'master' of github.com:daniperez/magrit

daniperez committed:
Unneeded argument, CMake already adds it.

daniperez committed:
Wrong description

daniperez committed:
Using version number given by maven

daniperez committed:
Version fixed (quotes removed)

daniperez committed:
CMake enabled.

Windows 32 bits cross-compilation in Fedora & Linux 32 bit compilation
should be working.

daniperez commented:
(Don't be obliged to merge this! I'm just seeking for your review so far)

ptitfred commented:
I'm pretty sure this doesn't work with maven.

Does the jar produced by `mvn client install` contains the library + the 2 executables ?

daniperez commented:
I only tested mvn check (or test? don't remember) in client folder. Can you point me to the error?

daniperez commented:
sorry, I understood your message it wrongly :-) my question was answered in your comment.

daniperez committed:
Rpath was missing

daniperez commented:
I tested it and it's working (I had just to fix the rpath in c2cf2c0afd9890ad9d24f5db6de15d49fdfcf164).

Nevertheless, the generated zip is weird: it contains both the bash scripts and mine executables. Was it the case before too?

ptitfred commented:
The zipping task is taking bash scripts and then override with yours, resulting a mix of shell scripts and binaries with the should-be-good result.

daniperez commented:
Sorry to tell that in my Fedora installation, the bash scripts overwrite the native executables :-(

ptitfred commented:
Ok I got it.

The packaging supposes the binaries produced from C++ are put in
/client/target/native-binaries. (Remember, the env var I call make with :P).

2012/3/8 daniperez <
reply@reply.github.com

> Sorry to tell that in my Fedora installation, the bash scripts overwrite
> the native executables :-(
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/ptitfred/magrit/pull/13#issuecomment-4401454

## 

Frdric Menou

daniperez commented:
I'm honoring that variable and in my system it's writing the binaries in /client/target/native-binaries. Isn't working in yours? where does it put the binaries?

daniperez committed:
Fixing some late-written readmes...

daniperez committed:
forgot this

daniperez committed:
fatal_error is better

ptitfred committed:
New output of build-log is light

ptitfred committed:
Merge branch 'robot' into for-dani

ptitfred committed:
Ignore CMake build directory

ptitfred committed:
Let the setup script accept other arguments

This is used by the robot to use a custom port (for instance to be launched in a Magrit server).

ptitfred committed:
T2: Test build-log magrit.log.maxwidth

ptitfred committed:
Custom port for testing Magrit instance

ptitfred committed:
Merge branch 'robot' into for-dani

daniperez committed:
Merge pull request #12 from ptitfred/for-dani

Merging ptitfred changes

daniperez committed:
Added .swo files to gitignore. Boost 1.40 is enough to compile the tool.

Boost 1.40 is the version shipped in Ubuntu Lucid... yet I need to
guess how to have a gcc 4.6 back-ported to that platform...

daniperez committed:
Fixes #11

daniperez committed:
Addresses #13.

daniperez committed:
Creating a proper library to avoid the symlink (not supported by the distribution archive).

daniperez committed:
Fixes #14.

daniperez committed:
Added test support.

In windows (wine), it's not finding the test suite because
the .exe extension o_O

I had to use a dynamically linked Boost for win32 because
otherwise I couldn't compile in both Linux and win32 with the
same flags. Maybe Boost dlls will be missing in the delivery.

daniperez committed:
After cross-compiling... cross-testing!

               __
              )==(
              )==(
              |H |
              |H |
              |H |
             /====\
            /      \
           /========\
          :HHHHHHHH H:
          |HHHHHHHH H|
          |HHHHHHHH H|
          |HHHHHHHH H|
   \______|=/========\________/
    \     :/oO/      |\      /
     \    / oOOO     | \    /
      \__/| OOO      |  \__/
       )( |  O       |   )(
       )( |==========|   )(
       )( |HHHHHHHH H|   )(
       )( |HHHHHHHH H|   )(
      .)(.|HHHHHHHH H|  .)(.
     ~~~~~~~~~~~~~~~~  ~~~~~~
 (powered by wine. Tests, not me)

daniperez committed:
Trying Travis with Magrit. We're good friends! (but don't merge this commit)

daniperez committed:
[travis test] added cmake to be installed

daniperez committed:
forgot to change directory

daniperez committed:
yet another error

daniperez committed:
yet another missing lib

daniperez committed:
[travis test] boost libs are not being found

daniperez committed:
[travis test] trying another package

daniperez committed:
[travis test] clean-up

daniperez committed:
Travis badge added

daniperez committed:
[travis test] started setting up configs

daniperez committed:
[travis test] trying windows config

daniperez committed:
[travis test] package clash fixed

daniperez committed:
[travis test] ubuntu support added

daniperez committed:
[travis test] trying to fix ubuntu config

daniperez committed:
Adding support to Ubuntu 11.10.

daniperez committed:
Kind of working in Ubuntu 11.10

daniperez committed:
Package not found

daniperez committed:
cmake -E tar doesn't work very well. Trying plain tar.

daniperez committed:
tar switch fixed

daniperez committed:
Using travis arch.

daniperez committed:
Travis is 64 bits and 32 bits not installed? Because it didn't work.

daniperez committed:
Using new mingw-boost packets in Github. Little progress.

daniperez committed:
Winepath fixed.

daniperez committed:
Ubuntu x86_64 cross-compilation for Win32 fixed.

daniperez committed:
These packages don't exist.

daniperez committed:
An EULA disclaimer related to MS fonts appears when I install deps.
I can't get rid of it because Travis works in non-interactive mode.
Trying to fix it.

daniperez committed:
ups, I put the preconfigure in a wrong place

daniperez committed:
Trying something else.

daniperez committed:
Yet not fixed... another try.

daniperez committed:
fix

daniperez committed:
It was reaching travis' timeout

daniperez committed:
Not installing boost-all, it eats up our Travis time.

daniperez committed:
typo

daniperez committed:
gcc links are missing

daniperez committed:
Fixing package not found

daniperez committed:
Trying different command lines

daniperez committed:
Using natty compiler naming

daniperez committed:
TRUE is not understood by cmake by default :-)

daniperez committed:
Trying with submodules

daniperez committed:
typo

daniperez committed:
Trying to fix Travis build

daniperez committed:
Every line is a config, only 2 configs needed.

daniperez committed:
tar fixed

daniperez committed:
yet another fix

daniperez committed:
Travis didn't like .cmake/mingw path

daniperez committed:
command switch order fixed

daniperez committed:
misunderstood git doc

daniperez committed:
...

daniperez committed:
trying with clone

daniperez committed:
forgot this

daniperez committed:
yet another try

daniperez committed:
commented windows config, ubuntu natty has an old compiler
